### ISSUE TITLE
Creating disulphide bond using tleap though the YAML file

### ISSUE THREAD

**Issue by mmagithub** [choderalab/yank#820]
Hi,
I am trying t use YANK in an automatic way to create a protein-ligand complex and run an absolute free energy calculation job.
I am asking if there is a way to add a disulphide bridge option in YAML file. I tried to modify the names to CYX and adding connect record to the PDB files of the receptors but none worked. Any suggestions?
Thanks
Marawan

**Comment by andrrizzi** [comment]
Hi Marawan.

> but none worked

Could you be a little more specific? Is the problem that the `pdb` in the `systems/system-name/complex.pdb` doesn't have any disulfide bond?

It would help if you included the YAML script you're using. Are you using the [`select`](http://getyank.org/0.18.0/yamlpages/molecules.html#select) keyword by any chance?

**Comment by mmagithub** [comment]
Hello Andrea,
Thanks for the reply.
Exactly, complex.pdb has no disulphide bond.
I also looked at the outraj.pdb file after the simulation finishes, no disulphide bond is there, i thought it may be only an artefact and the disulfide bond may exist in the parameter files created by tleap, but I could see it also. Furthermore, the CYS-SG-SG-CYS bond distance is high (> 3.5 Ang) is fluctuating a lot in the pdb trajectory (outraj.pdb) which indicates that the SG-SG bond order is zero. The complex_tleap.in file has no command to bond the two cysteines together, which is typically the case when you create a conventional tleap file by hand.  
No, I do not use the select command, do you think it will help?
My system is a small molecule bound receptor, so the disulfide bond should exist between residue 22 and 91 of the receptor file.
Here is the YAML script:
############
---
options:
  minimize: yes
  verbose: yes
  output_dir: .
  number_of_equilibration_iterations: 25
  number_of_iterations: 100
  nsteps_per_iteration: 500
  equilibration_timestep: 1.0*femtosecond
  temperature: 300*kelvin
  pressure: 1*atmosphere
  timestep: 1.5*femtoseconds
  output_dir: output
  resume_setup: yes
  resume_simulation: yes

molecules:
  REC:
    filepath: ./rec.pdb
  LIG:
    filepath: ./h_lig.mol2
    antechamber:
      charge_method: bcc

solvents:
  pme:
    nonbonded_method: PME
    switch_distance: 11*angstroms
    nonbonded_cutoff: 12*angstroms
    ewald_error_tolerance: 1.0e-4
    clearance: 16*angstroms
    positive_ion: Na+
    negative_ion: Cl-


systems:
  Rec-Lig:
    receptor: REC
    ligand: LIG
    solvent: pme
    leap:
      parameters: [leaprc.protein.ff14SB, leaprc.ff14SB,leaprc.gaff, leaprc.water.tip4pew,frcmod.tip4pew,frcmod.ionsjc_tip4pew]

protocols:
  absolute-binding:
    complex:
      alchemical_path:
        lambda_restraints:     [0.00, 0.05, 0.10, 0.15, 0.20, 0.25, 0.50, 0.75, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.000, 1.00, 1.000, 1.00, 1.00]
        lambda_electrostatics: [1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 0.95, 0.90, 0.85, 0.80, 0.75, 0.70, 0.65, 0.60, 0.55, 0.50, 0.45, 0.40, 0.35, 0.30, 0.25, 0.20, 0.15, 0.10, 0.05, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.000, 0.00, 0.000, 0.00, 0.00]
        lambda_sterics:        [1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 0.95, 0.90, 0.85, 0.80, 0.75, 0.70, 0.65, 0.60, 0.55, 0.50, 0.45, 0.40, 0.35, 0.30, 0.25, 0.20, 0.15, 0.10, 0.075, 0.05, 0.025, 0.01, 0.00]
    solvent:
      alchemical_path:
        lambda_electrostatics: [1.00, 0.95, 0.90, 0.85, 0.80, 0.75, 0.70, 0.65, 0.60, 0.55, 0.50, 0.45, 0.40, 0.35, 0.30, 0.25, 0.20, 0.15, 0.10, 0.05, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00]
        lambda_sterics:        [1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 0.95, 0.90, 0.85, 0.80, 0.75, 0.70, 0.65, 0.60, 0.55, 0.50, 0.45, 0.40, 0.35, 0.30, 0.25, 0.20, 0.15, 0.10, 0.00]

experiments:
  system: Rec-Lig
  protocol: absolute-binding
  restraint:
    type: Harmonic
############

BTW, the simulations finishes normally which means that the generated files are OK, and the CYX residues have dangling SG atoms (not protonated) and not bonded.

Any suggestion?

Regards,
Marawan

**Comment by mmagithub** [comment]
sorry, i meant "..but I could NOT see it also".

**Comment by andrrizzi** [comment]
Yeah, you are right. We don't currently support it. The quickest workaround now is to create the `prmtop`/`inpcrd` files systems yourself and start from those. Sorry about this.

For the future, could you wrap scripts between backticks? It really helps if the script is correctly indented. I've modified your post now.

Writing as a memo for later. To add support this we need to:
1. Load the receptor before running tleap and scan it for CYX-named residues.
2. Identify pairs of close CYX based on proximity (raise an exception when you have multiple cysteines within the threshold).
3. For each pair of CYX cysteines, add a line to the tleap script similar to bond `receptor.26.SG receptor.84.SG`.

**Comment by mmagithub** [comment]
Ok, thanks Andrea.
For the modification, I would rather recommend leaving this option to the user to either connect the CYX residues or not. Perhaps a new command within the YAML file to specify the residues that we need to create a disulphide bond for should be fine.
You never know if in certain circumstances one still needs to have a deprotonated CYS (conventionally named as CYX within AMBER) without a disulphide bond.
It is of course totally up to you, it is just a suggestion.
Another option is to allow YANK to deal with externally supplied tleap inputs. Or to create a new group of tleap related options under the name "tleap options" or something similar. This will add to the flexibility of YANK.
Have a nice day,
Regards,
Marawan

**Comment by andrrizzi** [comment]
> in certain circumstances one still needs to have a deprotonated CYS

Good point. I didn't think of that. An `extra_commands` keyword in the `tleap` section may be the best way to go about it then. I believe there's also some other issue where this idea had poped up. Thanks!

**Comment by mmagithub** [comment]
Great Andrea. Very soon people will start to use YANK with modified fragments, like lipid and sugar molecules, the power of tleap stems from its ability to work with these unusual systems. Giving YANK flexibility to deal with these systems through interfacing with tleap will enhance its widespread use..cannot wait to see these updates in actions..
Good luck,
Marawan

**Comment by andrrizzi** [comment]
Just writing a memo here that the conventional residue name for a deprotonated Cys that is not involved in a bond may be CYM instead of CYX.
